module5/lessons/Week4/ProcessOptimization.md

```diff
@@ -0,0 +1,246 @@
+
+When describing BigO, we `"n"` to signify the amount of data. The most common examples you might see include:
+
+- O(1)
+- O(log n)
+- O(n)
+- O(n log n)
+- O(n^2) (aka, n-squared)
+
+We'll discuss each of these in order from least complex (best performance) to most complex (worst performance).
+
+<section class="answer" markdown='1'>
+### Constant Time - O(1)
+
+**Description:**
+- No matter how much data there is, it only takes one operation.
+
+**Examples:**
+- Simple math operations and assignment operators
+- Invoking a function
+- Accessing an array by index position
+- Accessing a property in an object/hash by its attribute/key
+</section>
+
+<section class="answer" markdown='1'>
+### Logarithmic Time - O(log n)
+
+**Description:**
+- Every step/action you take, you cut out half of the work.
+
+**Examples:**
+- "Binary search" using a binary tree.  An example of doing this might include looking up a word in a dictionary.
+</section>
+
+<section class="answer" markdown='1'>
+### Linear Time - O(n)
+
+**Description:**
+- Complexity grows in direct proportion to the size of the input data.
+
+**Examples:**
+- Iteration over an array or the keys of an object/hash with a `for` loop or `while` loop
+- Using array prototype methods/enumerables such as `map` or `reduce`
+- Anything that looks at, or processes, every element in a data structure.
+</section>
+
+<section class="answer" markdown='1'>
+### Linearithmic Time - O(n log n)
+
+**Description:**
+- Often is a nested iteration where within each operation done in linear time, there are actions being done in logarithmic time over the same size of data
+
+**Examples:**
+- Efficient sorting algorithms such as `merge sort`
+- Some searching algorithms
+</section>
+
+<section class="answer" markdown='1'>
+### Quadratic Time - O(n^2)
+
+**Description:**
+- Nested loops over the same or similarly sized dataset
+
+**Examples:**
+- Some inefficient sorting algorithms such as `bubble sort`
+- Anytime you have nested loops or iteration methods
+</section>
+
+<section class="note" markdown='1'>
+
+Some Key Considerations
+
+Usually when calculating Big O notation, an algorithm doesn't fall nicely into one of the categories above. It's more likely that an algorithm combines some of the given strategies. Here are some key things to keep in mind:
+
+* When operations are sequential, their complexities add together
+* When operations are nested (within loops for example), their complexities multiply
+* When calculating Big O, we don't care about *constants* or *coefficients*. For example `O(2n + 1)` is simplified to `O(n)`
+* When calculating Big O, if your algorithm requires several sequential steps, your Big O notation would only list the worst of those complexities. For example, `O(n + n^2)` would be simplified to `O(n^2)`.
+
+</section>
+
```

Comment:
I think it would be helpful to have bit of practice in here, I think they are up for it and making things more concrete will help! Maybe have the two code blocks from earlier and then a couple more algorithms and they have to work with a partner to decide which efficiency this is.